src/Microsoft.DncEng.PatGeneratorTool/Program.cs

```diff
@@ -0,0 +1,169 @@
+                return 1;
+            }
+
+            if (string.IsNullOrEmpty(user) != string.IsNullOrEmpty(password))
+            {
+                Console.WriteLine("Must specify both user + password, or neither.");
+                return 1;
+            }
+
+            DateTime credentialExpiration = GetExpirationDate(expiration, expiresIn);
+
+            VssCredentials credentials;
+            if (!string.IsNullOrEmpty(user))
+            {
+                credentials = new VssAadCredential(user, password);
+            }
+            else
+            {
+                credentials = await GetInteractiveUserCredentials();
+            }
+
+            var patGenerator = new AzureDevOpsPATGenerator(credentials);
+            var pat = await patGenerator.GeneratePATAsync(patName, scopeFlags, organizations, credentialExpiration);
+            Console.WriteLine($"{patName} (Valid Until: {credentialExpiration}): {pat.Token}");
+            return 0;
+        }
+
+        /// <summary>
+        /// Determine the desired name of the PAT. It appears that PATs may have really any combination of characters
+        /// and they do not have to be unique.
+        /// </summary>
+        /// <param name="scopes">Desired scopes</param>
+        /// <param name="organizations">Organizations</param>
+        /// <param name="name">Name provided by the user</param>
+        /// <returns>Name of the PAT.</returns>
+        private static string GetPatName(AzureDevOpsPATScopes scopes, string[] organizations, string name)
+        {
+            string patName = name;
+            if (string.IsNullOrEmpty(patName))
+            {
+                string scopeString = scopes.GetScopeString();
+                patName = $"{string.Join("-", organizations)}-{scopeString}";
+            }
+
+            return patName;
+        }
+
+        /// <summary>
+        /// Compute the desired expiration date. If no expiration day or expiresIn is specified, 
+        /// then defaults to 30 days.
+        /// </summary>
+        /// <param name="expiration">Explicit expiration date</param>
+        /// <param name="expiresIn">Explicit number of days to expire in.</param>
+        /// <returns>Date that the credential should expire on.</returns>
+        private static DateTime GetExpirationDate(DateTime? expiration, int? expiresIn)
+        {
+            DateTime finalExpiration = DateTime.Now.AddDays(DefaultExpirationInDays);
+            if (!expiresIn.HasValue && expiration.HasValue)
+            {
+                finalExpiration = expiration.Value;
+            }
+            else if (expiresIn.HasValue && !expiration.HasValue)
+            {
+                finalExpiration = DateTime.Now.AddDays(expiresIn.Value);
+            }
+
+            return finalExpiration;
+        }
+
+        readonly static string[] AzureDevOpsAuthScopes = new string[] { "499b84ac-1321-427f-aa17-267ca6975798/user_impersonation" };
```

Comment:
Maybe a link to where this magic string comes from if it's somewhere reasonable?